### PR TITLE
7586 remove #ifdef __lint hack from dmu.h

### PR DESCRIPTION
**Pull request by dankimmel**
Reviewed by: Matthew Ahrens <mahrens@delphix.com>
Reviewed by: Pavel Zakharov <pavel.zakharov@delphix.com>

The `#ifdef __lint` in `dmu.h` is ugly, and it would be nice not to
duplicate it if we add other inline functions into header files in ZFS,
especially since it is difficult to make any other solution work across
all compilation targets. We should switch to disabling the lint flags
that are failing instead.

Upstream bugs: DLPX-44715

### CODE REVIEW

**Comment by dankimmel** [comment]
@zettabot go

File: usr/src/uts/common/fs/zfs/sys/dmu.h [hunk 1]
```diff
@@ -564,12 +564,7 @@ typedef struct dmu_buf_user {
  * NOTE: This function should only be called once on a given dmu_buf_user_t.
  *       To allow enforcement of this, dbu must already be zeroed on entry.
  */
-#ifdef __lint
-/* Very ugly, but it beats issuing suppression directives in many Makefiles. */
```
Comment (gwr) [inline usr/src/uts/common/fs/zfs/sys/dmu.h]: Is this really "better"?  You're trading away the ability to notice unused statics, which seems to me a high price to pay to get rid of an #ifdef __lint.

Comment (rmustacc) [inline usr/src/uts/common/fs/zfs/sys/dmu.h]: gcc will detect unused statics correctly but not get tripped up by inline functions. So it's true that lint won't notice, but as long a we aren't gagging unused functions, it should be fine.

Comment (gwr) [inline usr/src/uts/common/fs/zfs/sys/dmu.h]: If it's only lint that won't notice unused statics, then OK.

**Pull request closed without merging**